### PR TITLE
obs-filters: Warn when NVIDIA FX needs update, Update to v0.7.1 Video SDK & Add a mask refresh slider

### PR DESCRIPTION
**Pull request by pkviet**
### Description
This PR consists in three things:

1. Trying to mitigate an issue with NVIDIA Video & Audio FX sdks (which arises when updating one sdk and not the other).
2. Support v0.7.1.0+ Video FX sdk which improves Background Removal Filter by adding temporal processing.
3. Adding a mask refresh slider which controls how frequently the mask is refreshed.

### Motivation and Context

1. Context of the bug we are trying to mitigate: the NVIDIA Audio & Video Effects redistributable have been very recently updated (end of September 2022).
This can occasion bugs with obs-studio because the two redistributable rely on the same set of dependencies.
When one redistributable is more recent than the other, there can be a 'dll hell' with one of the two FX dll
being prevented from loading because the dependencies load but are the wrong ones for the other FX.
This is something which can not be solved on obs side but must be fixed by NVIDIA or the user.

- On the user side, the solution is to always upgrade both Video and Audio Effects redistributable at the same time. This is rendered a bit difficult because the versions of both SDKs have different numbers.
- On NVIDA side, an installer combining both Video & Audio FX would be one solution. Another would be to bundle the FX dlls with the GPU drivers.
On obs-studio side, we have implemented a detection of the dll versions and if they don't match a warning is issued in the Properties window of the filter.
Additionally the warning is logged.

2. Since Video Effects sdk v 0.7.0+, a major improvement is 'temporal processing' instead of a mere intraframe AI processing. This can improve the quality of the Background Removal filter. There is less flickering for instance.
We have added the capability to do temporal processing. But this will require the user to update to the latest sdk (as of Sept. 25). 
Back compatibility has been kept with earlier sdk should the user not want to upgrade.

3. The third feature (in the fourth commit) addresses one issue with the latest sdk (Video Effects sdk v 0.7.0+) which is much more demanding in terms of GPU processing than previous versions. For instance at 60 fps, obs can't always keep up if there's a demanding game (see also next section on tests for gory details) : there is rendering lag, which is detrimental. This is mitigated by leveraging the fact that it's not necessary to refresh the mask as frequently because usually we have a talking head who doesn't move too much. (This is an idea from Stephan Seitz, NVIDIA, whom I thank warmly.)
So we allow generating a mask every nth frame where n=1-4. For n=1, a mask is generated at each and every frame; for n=2, a mask is generated every two frames, etc ...
obs64_2022-10-25_00-11-21

### How Has This Been Tested?
This was tested with an RTX 3070 FE with both last SDK and previous one.
All the Audio and Video FXs work fine.
This was also tested with an RTX 4090.
The redistributables can be found [here](https://www.nvidia.com/en-us/geforce/broadcasting/broadcast-sdk/resources/).
The audio files are provided by NVIDIA in their sdks.
To demo the dereverb FX, I have dramatized the effect by inserting a reverb vst in obs.

1. Audio FX:
   a. Denoiser (NVIDIA demo file played in obs + noise reduction filter with NVIDIA FX set to noise reduction): 

https://user-images.githubusercontent.com/9283407/192160933-8c1602dd-d308-4590-8318-4b24ae9efb5c.mp4

   b. Dereverb (NVIDIA demo file played in obs + reverb vst + noise reduction filter with NVIDIA FX set to dereverb): 

https://user-images.githubusercontent.com/9283407/192160936-52951685-e1b3-48a9-ba94-583ccc7a80b6.mp4

   c. Dereverb + Denoiser (NVIDIA demo file played in obs + reverb vst + noise reduction filter with NVIDIA FX set to dereverb+denoiser): 

https://user-images.githubusercontent.com/9283407/192160920-29d6e4a0-77af-402c-af9d-9d0203ae44a4.mp4

2. Video FX (Background Removal filter):

We checked that the filter still works after the changes in this PR with various types of sources.

- Browser-source:
obs64_2022-09-25_20-35-26

- Video source:
obs64_2022-09-25_20-40-52

- Still image:
obs64_2022-09-25_20-35-40

- webcam (not shown).

3. Video FX (Background Removal filter) benchmark:
- I made extensive benchmarking of the effect by having obs and cyberpunk 2077 benchmark. Each figure in the table below corresponds to 5-10 runs of cyberpunk benchmark.
- hardware: RTX 4090 , 64 GB DDR4 , CPU: 5950x
- obs video settings: 1080p60 (720p60, 2k and 4k show similar results).

|   Setup | cyberpunk alone  | obs 28.0.3 | obs 28.0.3 + nv broadcast  |  obs 28.0.3 + fx quality  |  obs 28.0.3 + fx perf |  PR + fx quality  | PR + fx quality + skip 1 mask   | PR + fx quality + skip 2 masks   | PR + fx quality + skip 3 masks  | PR + fx perf  | PR + fx perf + skip 1 mask  |  PR + fx perf + skip 2 masks | 
|---|---|---|---|---|---|---|---|---|---|---|---|---| 
|  FPS | 141.0±0.3  | 125.2±0.3  | 113.7±0.2  | 105.7±0.2  | 105.0±0.4  | 110.6±0.2 (obs: rendering lag)  | 119.3±0.2  | 123.8±0.1  | 125.5±0.3  | 114.2±0.1 (obs: rendering lag)   | 118.7±0.1  | 123.1±0.3  |  

The various setups:
- obs 28.0.3 + nv broadcast : cyberpunk benchmark + current release of obs + webcam video source using nvidia virtual webcam with background removal from NVIDIA Broadcast app.
- obs 28.0.3 + fx quality : current release of obs + webcam video source + background removal filter in Quality mode (better quality but higher gpu load).
- obs 28.0.3 + fx perf : current release of obs + webcam video source + background removal filter in Perf mode (good quality and lower gpu load).
- PR + fx quality : obs with current PR + webcam video source + background removal filter in Quality mode.
- PR + fx quality + skip 1 mask : obs with current PR + webcam video source + background removal filter in Quality mode + mask generated every two frames (one skips 1 frame between mask generation).
- PR + fx quality + skip 2 masks : obs with current PR + webcam video source + background removal filter in Quality mode + mask generated every three frames (one skips 2 frames between mask generation).
- PR + fx quality + skip 3 masks : obs with current PR + webcam video source + background removal filter in Quality mode + mask generated every four frames (one skips 3 frames between mask generation).
- PR + fx perf : obs with current PR + webcam video source + background removal filter in Perf mode.
- PR + fx perf + skip 1 mask : obs with current PR + webcam video source + background removal filter in Perf mode + mask generated every two frames (one skips 1 frame between mask generation).
- PR + fx perf + skip 2 masks : obs with current PR + webcam video source + background removal filter in Perf mode + mask generated every three frames (one skips 2 frames between mask generation).

Observations:
- one sees that Quality and Perf modes don't make much difference in terms of GPU load.
- the GPU load is higher with the PR which enables temporal processing.
- one finds that with the PR sdk update, obs starts to have some rendering lag and can't keep at 60 fps.
- this rendering lag disappears though if masks are not generated at every frame.
- the native integration of NVIDIA Background removal is more efficient than using NVIDIA Broadcast app if one allows mask generation every two frames (30 masks/sec at 60 fps). 

### Types of changes
- Bug fix (non-breaking change which fixes an issue) 
- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by pkviet** [comment]
Fixes issue #7369

File: plugins/obs-filters/nvidia-greenscreen-filter.c [hunk 1]
```diff
@@ -68,6 +75,13 @@ struct nv_greenscreen_data {
 	gs_eparam_t *threshold_param;
 	gs_eparam_t *multiplier_param;
 	float threshold;
+
+	/* Every nth frame is processed through the FX (where n = partial_processing)
+     * so that the same mask is used for n consecutive frames. 
+     * This is to alleviate the GPU load. Default: 1.
+     */
```
Comment (RytoEX) [inline plugins/obs-filters/nvidia-greenscreen-filter.c:82]: Nit: Fix indentation.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Just a nit.

File: plugins/obs-filters/nvidia-greenscreen-filter.c (hunk 1 above)
Comment (pkviet) [reply plugins/obs-filters/nvidia-greenscreen-filter.c]: fixed

**Review by Fenrirthviti — APPROVED** [review]

**Review by RytoEX — APPROVED** [review]

**Comment by sandr1x** [comment]
Do I understand correctly that Nvidia Audio SDK is not necessary unless new features are needed?

**Comment by khaosmaou** [comment]
This warning message shows up on 10XX series cards and older. There are no Audio SDK updates from Nvidia for cards of these generations. The warning message indicates something is wrong, but as RNNoise still works on my GTX 1080 with the latest drivers with no issue this message can be misleading. Installing the new SDK adds new options and makes the warning message disappear, but none of these new "features" are even supported on non-RTX cards and show as greyed out inside of OBS.

Seems like this warning should only show on 20XX series or newer cards, and only if there is an update available to the SDK to prevent users with older cards installing extra software that literally does nothing.

**Comment by max-programming** [comment]
I have a GTX 1650 but this warning shows up and also the Noise Suppression filter simply just **doesn't work** It doesn't remove noise from my background as it did before

**Pull request merged**